Symbol type (#66)
javascript-tutorial/zh-hant.javascript.info: 1-js/04-object-basics/03-symbol/article.md

user[id] = "Their id value";
```

> 這不會造成我們與其他腳本之間有任何衝突，因為 Symbol 永遠是不同的，即使他們有相同的名稱。

Reviewer: @ArvinH how about this?

> 這不會造成我們與其他腳本之間有任何衝突，因為 Symbol 永遠是不同的，即使他們有相同的名稱。

Reviewer: 我發現這在你的另一個 PR 有改動過，所以被標成 outdated，但 suggestion 的部分沒改

> 這不會造成我們與其他腳本之間有任何衝突，因為 Symbol 永遠是不同的，即使它們有相同的名稱。

Reviewer: ```suggestion
這不會造成我們與其他腳本之間的識別符有任何衝突，因為 Symbol 永遠是不同的，即使它們有相同的名稱。
```

> 這不會造成我們與其他腳本之間有任何衝突，因為 Symbol 永遠是不同的，即使他們有相同的名稱。

Reviewer: 我在下面開一個新的

alert( clone[id] ); // 123
```

這裡沒有悖論。就是這樣設計的。想法是當我們複製一個物件，或是合併多個物件時，我們通常想要 *所有* 屬性都被複製（包含像 `id` 這樣的 Symbol ）。

> ````smart header="其他類別的屬性件值會被強行轉換成字串"

Author: ```suggestion
````smart header="其他類別的屬性鍵值會被強行轉換成字串"
```

在物件中，我們只能使用字串或 Symbol 當作鍵值。其他類型都會被轉成字串。

舉例來說，當一個數字 `0` 被用來當作屬性健值時，它會變成一個字串 `"0"`：

```js run
let obj = {
  0: "test" // 跟 "0": "test" 一樣
};

// 兩個 alerts 都存取到同樣的屬性（數字 0 被轉換成字串 "0"）
alert( obj["0"] ); // test
alert( obj[0] ); // test (同樣的屬性)
```
````

## 全局符號

正如我們所見，通常所有的 Symbol 都是不同的，即使他們擁有相同的名稱。但是有時候我們想要擁有相同名稱的 Symbol 被當作相同的物體。例如，我們應用程式中的不同部分想用Symbol `"id"` 存取到完全相同的屬性。

> 為此，存在一個 *全局 Symbol 註冊表*。我們可以在其中創建 Symbol 並在稍後存取他們，而這確保我們每次存取相同名稱都會回傳相同的 Symbol。

Author: ```suggestion
為此，存在一個 *全域 Symbol 註冊表（global symbol registry）*。我們可以在其中創建 Symbol 並在稍後存取他們，而這確保我們每次存取相同名稱都會回傳相同的 Symbol。
```

為了從註冊表中讀取（如果不存在就創建）Symbol，請使用 `Symbol.for(key)`。

該呼叫會檢查全局註冊表，若有個描述為 `key` 的 Symbol 存在，則回傳該 Symbol，否則就創建新的 Symbol `Symbol(key)` 並將其以 `key` 儲存到註冊表中。

舉個例子：

```js run
// 從全局註冊表中讀取
let id = Symbol.for("id"); // 如果該 Symbol 不存在，就創建它

// 再次讀取它（或許從程式碼的另一個部分）
let idAgain = Symbol.for("id");

// 同樣的 Symbol
alert( id === idAgain ); // true
```

註冊表內的 Symbols 稱為 *全局 Symbol*。如果我們想要一個應用程序範圍內的 Symbol，可以在程式碼中隨處存取 -- 這就是它們的用途。

```smart header="這聽起來像是 Ruby"
在某些程式語言中，像是 Ruby，每個 Symbol 都有個名稱。

> 在 JavaScript 中，如我們所見，只有全局 Symbol 才是如此。

Author: lol 難怪覺得怪怪的  一時想不起來 全域 這個詞

```

### Symbol.keyFor

> 對於全局 Symbol，不止有 `Symbol.for(key)` 根據名稱回傳 Symbol，還有個反向呼叫：`Symbol.keyFor(sym)`，反過來，他根據全局 Symbol 回傳名稱。

Author: ```suggestion
對於全域 Symbol，不止有 `Symbol.for(key)` 可以根據某個名稱回傳其 Symbol，還有個反向呼叫，`Symbol.keyFor(sym)` 做反過來的事：根據一個全域 Symbol 回傳其名稱。
```

例如：

```js run
// 根據名稱取得 Symbol
let sym = Symbol.for("name");
let sym2 = Symbol.for("id");

// 根據名稱取得 Symbol
alert( Symbol.keyFor(sym) ); // name
alert( Symbol.keyFor(sym2) ); // id
```

`Symbol.keyFor` 在內部使用全局 Symbol 註冊表來查找 Symbol 的鍵值。所以它並不適用於非全局 Symbol。如果有一個 Symbol 不是全局的，此方法將無法找到它，並會回傳 `undefined`。

是說，任何 Symbol 都有 `description` 屬性。

例如：

```js run
let globalSymbol = Symbol.for("name");
let localSymbol = Symbol("name");

alert( Symbol.keyFor(globalSymbol) ); // name, 全局 symbol
alert( Symbol.keyFor(localSymbol) ); // undefined, 非全局

alert( localSymbol.description ); // name
```

## 系統符號

JavaScript 內部存在許多 "系統" Symbol，我們可以使用它們來微調物件的各種方面。

> 它們列在[熟悉的 symbols](https://tc39.github.io/ecma262/#sec-well-known-symbols) 表的規格中：

Reviewer: spec 的特定某章名稱好像可以不用翻？
```suggestion
它們列在規格中的 [Well-known symbols](https://tc39.github.io/ecma262/#sec-well-known-symbols) 表格內：
```

- `Symbol.hasInstance`
- `Symbol.isConcatSpreadable`
- `Symbol.iterator`
- `Symbol.toPrimitive`
- ...等等。

> 例如，`Symbol.toPrimitive` 允許我們將物件描述為原始值轉換。我們很快就會看到它的使用。

Reviewer: `primitive`  我前面都翻 `原生` 或 `原生值`，可以統一一下～
```suggestion
例如，`Symbol.toPrimitive` 允許我們將物件描述為原生值轉換。我們很快就會看到怎麼使用它。
```

當我們研讀相應的語言特性時，也將會更熟悉其他 Symbol。

## 總結

> `Symbol` 是唯一識別符號的原始類別。

Reviewer: ```suggestion
`Symbol` 是用於標示唯一識別符號的原生類別。
```

> Symbol 使用 `Symbol()` 與一個可選的敘述作為參數來創建。

Reviewer: ```suggestion
Symbol 使用 `Symbol()` 與一個可選的敘述（名稱）作為參數來創建。
```

> Symbol 永遠是不同的值，即使他們擁有相同的名稱。如果我們想要同名的 Symbol 也相等，那我們應該使用全局註冊表：`Symbol.for(key)` 回傳（如果需要的話創建）一個以 `key` 作為名稱的全局 Symbol。針對相同的 `key` 以 `Symbol.for` 進行多次呼叫，都會回傳相通的 Symbol。

Reviewer: ```suggestion
Symbol 永遠是不同的值，即使他們擁有相同的名稱。如果我們想要同名的 Symbol 也相等，那我們應該使用全域註冊表：`Symbol.for(key)` 回傳（如果需要的話創建）一個以 `key` 作為名稱的全域 Symbol。以相同的 `key` 用 `Symbol.for` 多次進行呼叫，都會回傳完全相同的 Symbol。
```

> Symbol 永遠是不同的值，即使他們擁有相同的名稱。如果我們想要同名的 Symbol 也相等，那我們應該使用全局註冊表：`Symbol.for(key)` 回傳（如果需要的話創建）一個以 `key` 作為名稱的全局 Symbol。針對相同的 `key` 以 `Symbol.for` 進行多次呼叫，都會回傳相通的 Symbol。

Author: ```suggestion
Symbol 永遠是不同的值，即使他們擁有相同的名稱。如果我們想要同名的 Symbol 也相等，那我們應該使用全域註冊表：`Symbol.for(key)` 回傳（如果需要的話創建）一個以 `key` 作為名稱的全域 Symbol。以相同的 `key` 用 `Symbol.for` 多次進行呼叫，都會回傳完全相同的 Symbol。
```

Symbol 有兩個主要使用場景：

1. "隱藏" 物件屬性。

>     如果我們想要增加一個屬性到屬於其他腳本或是套件的物件之中，我們可以創建一個 Symbol 並用它當作屬性的鍵值。Symbol 屬性不會出現在 `for..in` 中，所以他不會不小心被其他屬性一起處理。此外它也不能被直接存取，因為其他腳本不擁有我們的 Symbol。所以該屬性將會被保護，以防意外被存取或覆寫。

Reviewer: ```suggestion
    如果我們想要增加一個屬性到 "屬於" 其他腳本或是套件的物件之中，我們可以創建一個 Symbol 並用它當作屬性的鍵值。Symbol 屬性不會出現在 `for..in` 中，所以他不會不小心被其他屬性一起處理。此外它也不能被直接存取，因為其他腳本不擁有我們的 Symbol。所以該屬性將會被保護，以防意外被存取或覆寫。
```

>     如果我們想要增加一個屬性到屬於其他腳本或是套件的物件之中，我們可以創建一個 Symbol 並用它當作屬性的鍵值。Symbol 屬性不會出現在 `for..in` 中，所以他不會不小心被其他屬性一起處理。此外它也不能被直接存取，因為其他腳本不擁有我們的 Symbol。所以該屬性將會被保護，以防意外被存取或覆寫。

Author: ```suggestion
    如果我們想要增加一個屬性到 "屬於" 其他腳本或是套件的物件之中，我們可以創建一個 Symbol 並用它當作屬性的鍵值。Symbol 屬性不會出現在 `for..in` 中，所以他不會不小心被其他屬性一起處理。此外它也不能被直接存取，因為其他腳本不擁有我們的 Symbol。所以該屬性將會被保護，以防意外被存取或覆寫。
```

>     所以我們可以使用 Symbol 屬性， "秘密地" 將一些我們需要，但其他人不需要的東西藏進物件中。

Reviewer: 感覺不需要這個逗號？
```suggestion
    所以我們可以使用 Symbol 屬性， "秘密地" 將一些我們需要但其他人不需要的東西藏進物件中。
```

>     所以我們可以使用 Symbol 屬性， "秘密地" 將一些我們需要，但其他人不需要的東西藏進物件中。

Author: ```suggestion
    所以我們可以使用 Symbol 屬性， "秘密地" 將一些我們需要，但其他人不需要的東西藏進物件中。
```

> 2. JavaScript 內部使用了許多系統 Symbol，這些 Symbol 可以以 `Symbol.*` 被存取。我們可以使用它們來更改一些內建的行為。舉例來說，在後面的教學我們將會在 [迭代](info:iterable) 中使用 `Symbol.iterator`，設定 [物件轉換原始值](info:object-toprimitive) 時使用 `Symbol.toPrimitive`，等等

Reviewer: 我看沒有 `Symbol.*` 這個語法耶，應該是想表示這種形式而已

```suggestion
2. JavaScript 內部使用了許多系統 Symbol，以類似 `Symbol.*` 的形式被存取。我們可以使用它們來更改一些內建的行為。舉例來說，在之後的教程我們將會在 [迭代（iterables）](info:iterable) 中使用 `Symbol.iterator`，設定 [物件轉換原生值（object-to-primitive conversion）](info:object-toprimitive) 時使用 `Symbol.toPrimitive`，等等。
```

為了從註冊表中讀取（如果不存在就創建）Symbol，請使用 `Symbol.for(key)`。

該呼叫會檢查全域註冊表，若有個描述為 `key` 的 Symbol 存在，則回傳該 Symbol，否則就以 `Symbol(key)` 創建新的 Symbol，並將其以 `key` 儲存到註冊表中。

舉個例子：

```js run
// 從全域註冊表中讀取
let id = Symbol.for("id"); // 如果該 Symbol 不存在，就創建它

// 再次讀取它（或許從程式碼的另一個部分）
let idAgain = Symbol.for("id");

// 同樣的 Symbol
alert( id === idAgain ); // true
```

註冊表內的 Symbols 稱為 *全域 Symbol*。如果我們想要一個應用程序範圍內的 Symbol，可以在程式碼中隨處存取 -- 這就是它們的用途。

```smart header="這聽起來像是 Ruby"
在某些程式語言中，像是 Ruby，每個名稱都只有單一個 symbol。

在 JavaScript 中，如我們所見，只有全域 Symbol 才是如此。
```

### Symbol.keyFor

對於全域 Symbol，不止有 `Symbol.for(key)` 可以根據某個名稱回傳其 Symbol，還有個反向呼叫，`Symbol.keyFor(sym)` 做反過來的事：根據一個全域 Symbol 回傳其名稱。

例如：

```js run
// 根據名稱取得 Symbol
let sym = Symbol.for("name");
let sym2 = Symbol.for("id");

// 根據 Symbol 取得名稱
alert( Symbol.keyFor(sym) ); // name
alert( Symbol.keyFor(sym2) ); // id
```

`Symbol.keyFor` 在內部使用全域 Symbol 註冊表來查詢 Symbol 的鍵值。所以它並不適用於非全域的 Symbol。如果某個 Symbol 不是全域的，此方法將無法找到它，並會回傳 `undefined`。

是說，任何 Symbol 都有 `description` 屬性。

例如：

```js run
let globalSymbol = Symbol.for("name");
let localSymbol = Symbol("name");

alert( Symbol.keyFor(globalSymbol) ); // name, 全局 symbol

> alert( Symbol.keyFor(globalSymbol) ); // name, 全域 symbol

Reviewer: 欸靠北我發現 multiline suggestion 還是不行，來幹掉一行：
```suggestion
```
